Source/Core/DolphinQt/QtUtils/RunOnObject.h

```diff
@@ -3,74 +3,34 @@
 
 #pragma once
 
-#include <QCoreApplication>
-#include <QEvent>
-#include <QPointer>
-#include <QThread>
-#include <optional>
-#include <type_traits>
-#include <utility>
+#include <QMetaObject>
 
-#include "Common/Event.h"
-#include "DolphinQt/QtUtils/QueueOnObject.h"
+#include <concepts>
+#include <functional>
+#include <optional>
 
-class QObject;
+#include "Common/OneShotEvent.h"
 
 // QWidget and subclasses are not thread-safe! This helper takes arbitrary code from any thread,
 // safely runs it on the appropriate GUI thread, waits for it to finish, and returns the result.
 //
-// If the target object is destructed before the code gets to run, the QPointer will be nulled and
-// the function will return nullopt.
+// If the target object is destructed before the code gets to run the function will return nullopt.
 
-template <typename F>
-auto RunOnObject(QObject* object, F&& functor)
+auto RunOnObject(QObject* object, std::invocable<> auto&& functor)
 {
-  using OptionalResultT = std::optional<std::invoke_result_t<F>>;
-
-  // If we queue up a functor on the current thread, it won't run until we return to the event loop,
-  // which means waiting for it to finish will never complete. Instead, run it immediately.
-  if (object->thread() == QThread::currentThread())
-    return OptionalResultT(functor());
-
-  class FnInvokeEvent : public QEvent
-  {
-  public:
-    FnInvokeEvent(F&& functor, QObject* obj, Common::Event& event, OptionalResultT& result)
-        : QEvent(QEvent::None), m_func(std::move(functor)), m_obj(obj), m_event(event),
-          m_result(result)
-    {
-    }
-
-    ~FnInvokeEvent()
-    {
-      if (m_obj)
-      {
-        m_result = m_func();
-      }
-      else
-      {
-        // is already nullopt
-      }
-      m_event.Set();
-    }
+  Common::OneShotEvent task_complete;
+  std::optional<decltype(functor())> result;
 
-  private:
-    F m_func;
-    QPointer<QObject> m_obj;
-    Common::Event& m_event;
-    OptionalResultT& m_result;
-  };
+  QMetaObject::invokeMethod(
+      object, [&, guard = Common::ScopedSetter{&task_complete}] { result = functor(); });
```

Comment:
Is using `ScopedSetter` (or `ScopeGuard` in earlier revisions of this PR) safer in some way than just calling `task_complete.Set()` directly?